ui/src/app/base/sagas/websockets.ts

```diff
@@ -0,0 +1,627 @@
+};
+export const deleteFileContextRequest = (
+  id: WebSocketRequest["request_id"]
+): void => {
+  fileContextRequests.delete(id);
+};
+
+/**
+ * Whether the data is fetching or has been fetched into state.
+ *
+ * @param endpoint - root redux model and method (e.g. 'users.list')
+ * @returns - has data been fetched?
+ */
+const isLoaded = (endpoint: WebSocketEndpoint) => {
+  return loadedEndpoints.includes(endpoint);
+};
+
+/**
+ * Mark a model as having been fetched into state.
+ *
+ * @param endpoint - root redux state model and method (e.g. 'users.list')
+ */
+const setLoaded = (endpoint: WebSocketEndpoint) => {
+  if (!isLoaded(endpoint)) {
+    loadedEndpoints.push(endpoint);
+  }
+};
+/**
+ * Reset the list of loaded endpoints.
+ *
+ */
+const resetLoaded = () => {
+  loadedEndpoints = [];
+};
+
+/**
+ * Create a WebSocket connection via the client.
+ */
+export function createConnection(
+  websocketClient: WebSocketClient
+): Promise<WebSocketClient> {
+  // As the socket automatically tries to reconnect we don't reject this
+  // promise, but rather wait for it to eventually connect.
+  return new Promise((resolve, reject) => {
+    const readyState = websocketClient.socket?.readyState;
+    if (readyState === WebSocket.OPEN) {
+      resolve(websocketClient);
+      return;
+    } else if (
+      !websocketClient.socket ||
+      (readyState && [WebSocket.CLOSED, WebSocket.CLOSING].includes(readyState))
+    ) {
+      try {
+        // Check that the csrftoken etc. exist to create the connection. The
+        // check is done here because we don't want to reject errors when
+        // connecting so that the reconnecting websocket can keep trying.
+        websocketClient.buildURL();
+      } catch (error) {
+        reject(error);
+      }
+      websocketClient.connect();
+    }
+    if (websocketClient.socket) {
+      websocketClient.socket.onopen = () => {
+        resolve(websocketClient);
+      };
+    }
+  });
+}
+
+/**
+ * Create a channel to handle WebSocket messages.
+ */
+export function watchMessages(socketClient: WebSocketClient): WebSocketChannel {
+  return eventChannel((emit) => {
+    if (socketClient.socket) {
+      socketClient.socket.onmessage = (event) => {
+        emit(event);
+      };
+      socketClient.socket.onopen = (event) => {
+        emit(event);
+      };
+      socketClient.socket.onerror = (event) => {
+        emit(event);
+      };
+      socketClient.socket.onclose = (event) => {
+        emit(event);
+      };
+    }
+    return () => {
+      socketClient.socket?.close();
+    };
+  });
+}
+
+/**
+ * Handle incoming notify messages.
+ *
+ * Notify messages have an action and a payload:
+ * {"type": 2,
+ *  "name": "config",
+ *  "action": "update",
+ *  "data": {"name": "maas_name", "value": "maas-hysteria"}}
+ *
+ * Although we receive a corresponding response for each websocket requests,
+ * the store is only updated once a notify message has been received.
```

Comment:
This isn't always true. Maybe "in some cases the store is only..."

Reply:
Done.